src/jdk.dynalink/share/classes/jdk/dynalink/BiClassValue.java

```diff
@@ -96,33 +102,47 @@
             }
         }
 
-        private volatile Map<Class<?>, T> forward;
-        private volatile Map<Class<?>, T> reverse;
+        private volatile Map<Class<?>, T> forward = Map.of();
+        private volatile Map<Class<?>, T> reverse = Map.of();
 
         T getForwardValue(final Class<?> c) {
-            return getValue(forward, c);
+            return forward.get(c);
         }
 
         T getReverseValue(final Class<?> c) {
-            return getValue(reverse, c);
+            return reverse.get(c);
         }
 
-        private static <T> T getValue(Map<Class<?>, T> m, final Class<?> c) {
-            return m != null ? m.get(c) : null;
+        T compute(Map<Class<?>, T> map, final VarHandle mapHandle, final Class<?> c, final Function<Class<?>, T> compute) {
+            if (!map.containsKey(c)) {
+                final T value = compute.apply(c);
+                if (value == null) {
+                    return null;
+                }
+                do {
+                    final var entries = new ArrayList<>(map.entrySet());
+                    entries.add(Map.entry(c, value));
+                    @SuppressWarnings("rawtypes")
+                    final var newEntries = entries.toArray(new Map.Entry[0]);
```

Comment:
Since map is immutable now, instead of:

```
var entries = new ArrayList<>(map.entrySet());
entries.add(Map.entry(c, value));
var newEntries = entries.toArray(new Map.Entry[0]); 
```

you could write:

```
var entries = map.entrySet().toArray(new Map.Entry[map.size() + 1]);
entries[map.size()] = Map.entry(c, value);
```